README.md
# FXデータ・クローラー _(investment-fx-crawler)_

[![Build Status](https://travis-ci.org/u6k/investment-fx-crawler.svg?branch=master)](https://travis-ci.org/u6k/investment-fx-crawler) [![GitHub release](https://img.shields.io/github/release/u6k/investment-fx-crawler)](https://github.com/u6k/investment-fx-crawler/releases) [![GitHub](https://img.shields.io/github/license/u6k/investment-fx-crawler)](https://github.com/u6k/investment-fx-crawler/blob/master/LICENSE) [![Website](https://img.shields.io/website/https/redmine.u6k.me?up_message=redmine)](https://redmine.u6k.me/projects/investment-fx-crawler)

> FXのデータを収集する

__Table of Contents__

- [Install](#Install)
- [Usage](#Usage)
- [Other](#Other)
- [Disclaimer](#Disclaimer)
- [Maintainer](#Maintainer)
- [Contributing](#Contributing)
- [License](#License)

## Install

__TODO:__ 書く

## Usage

__TODO:__ 書く

## Other

その他の情報は、[Wiki](https://redmine.u6k.me/projects/investment-fx-crawler/wiki/Wiki/edit)を参照してください。

## Disclaimer

- 本プロジェクトの内容は、作者個人が趣味の範囲で自発的に行ったものです。作者が所属する会社または団体は本プロジェクトについて、一切関知しておらず、何ら責任や関連はありません。
- 本プロジェクトの内容は、作者個人が運用する目的としたものであり、投資の勧誘を目的としたものではありません。
- 本プロジェクトの内容またはそれを元にして作成したアプリケーションあるいはそれによる予測(以下、その成果物が生成する情報と呼称する)を参考にして、投資の判断を行わないでください。また、本プロジェクトの内容またはその成果物が生成する情報が正確であることを、作者は一切保証しません。未来の出来事に対する予想も、あくまで予想であり、その時々の状況によって変わりうることにご注意ください。
- 投資を行う際には、投資を行う本人のみの判断と責任において行ってください。
- 仮に投資に失敗し、投資を行ったものの資産が減ったり債務を負ったりしても、作者は何ら責任を負いません。
- その他、本プロジェクトを参考にしたかどうかに関わらず、投資の失敗や機会の損失などを含むいかなる不利益が何人に生じたとしても、作者は何ら責任を負いません。

## Maintainer

- u6k
    - [Twitter](https://twitter.com/u6k_yu1)
    - [GitHub](https://github.com/u6k)
    - [Blog](https://blog.u6k.me/)

## Contributing

当プロジェクトに興味を持っていただき、ありがとうございます。素晴らしいアイデアをお持ちでしたら、[新しいチケットを起票](https://redmine.u6k.me/projects/investment-fx-crawler/issues)していただけると幸いです。

当プロジェクトは、[Contributor Covenant](https://www.contributor-covenant.org/version/1/4/code-of-conduct)に準拠します。

## License

[MIT License](https://github.com/u6k/investment-fx-crawler/blob/master/LICENSE)



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.0] - 2019-08-19

### Changed

- [#7279: データを10秒足、1時間足、1日足のみにする](https://redmine.u6k.me/issues/7279)

## [1.1.0] - 2019-08-18

### Changed

- [#7272: 通貨ペア過去履歴データから、1秒足データ、10秒足データ、1分足データを生成する](https://redmine.u6k.me/issues/7272)

## [1.0.0] - 2019-08-17

### Added

- [#7271: 主要通貨ペアの過去履歴データを取得して、DBに格納する](https://redmine.u6k.me/issues/7271)



histdata_scrapy/settings.py
# -*- coding: utf-8 -*-

# Scrapy settings for histdata_scrapy project
#
# For simplicity, this file contains only settings considered important or
# commonly used. You can find more settings consulting the documentation:
#
#     https://docs.scrapy.org/en/latest/topics/settings.html
#     https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
#     https://docs.scrapy.org/en/latest/topics/spider-middleware.html

BOT_NAME = 'histdata_scrapy'

SPIDER_MODULES = ['histdata_scrapy.spiders']
NEWSPIDER_MODULE = 'histdata_scrapy.spiders'


# Crawl responsibly by identifying yourself (and your website) on the user-agent
# USER_AGENT = 'histdata_scrapy (+http://www.yourdomain.com)'
USER_AGENT = 'Mozilla/5.0 (X11; CrOS x86_64 6415.2.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2202.5 Safari/537.36'

# Obey robots.txt rules
ROBOTSTXT_OBEY = False

# Configure maximum concurrent requests performed by Scrapy (default: 16)
# CONCURRENT_REQUESTS = 32
CONCURRENT_REQUESTS = 1

# Configure a delay for requests for the same website (default: 0)
# See https://docs.scrapy.org/en/latest/topics/settings.html#download-delay
# See also autothrottle settings and docs
DOWNLOAD_DELAY = 1
# The download delay setting will honor only one of:
# CONCURRENT_REQUESTS_PER_DOMAIN = 16
# CONCURRENT_REQUESTS_PER_IP = 16

# Disable cookies (enabled by default)
# COOKIES_ENABLED = False

# Disable Telnet Console (enabled by default)
# TELNETCONSOLE_ENABLED = False

# Override the default request headers:
# DEFAULT_REQUEST_HEADERS = {
#   'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
#   'Accept-Language': 'en',
# }

# Enable or disable spider middlewares
# See https://docs.scrapy.org/en/latest/topics/spider-middleware.html
# SPIDER_MIDDLEWARES = {
#    'histdata_scrapy.middlewares.HistdataScrapySpiderMiddleware': 543,
# }

# Enable or disable downloader middlewares
# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
# DOWNLOADER_MIDDLEWARES = {
#    'histdata_scrapy.middlewares.HistdataScrapyDownloaderMiddleware': 543,
# }

# Enable or disable extensions
# See https://docs.scrapy.org/en/latest/topics/extensions.html
# EXTENSIONS = {
#    'scrapy.extensions.telnet.TelnetConsole': None,
# }

# Configure item pipelines
# See https://docs.scrapy.org/en/latest/topics/item-pipeline.html
# ITEM_PIPELINES = {
#    'histdata_scrapy.pipelines.HistdataScrapyPipeline': 300,
# }

# Enable and configure the AutoThrottle extension (disabled by default)
# See https://docs.scrapy.org/en/latest/topics/autothrottle.html
# AUTOTHROTTLE_ENABLED = True
# The initial download delay
# AUTOTHROTTLE_START_DELAY = 5
# The maximum download delay to be set in case of high latencies
# AUTOTHROTTLE_MAX_DELAY = 60
# The average number of requests Scrapy should be sending in parallel to
# each remote server
# AUTOTHROTTLE_TARGET_CONCURRENCY = 1.0
# Enable showing throttling stats for every response received:
# AUTOTHROTTLE_DEBUG = False

# Enable and configure HTTP caching (disabled by default)
# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html#httpcache-middleware-settings
HTTPCACHE_ENABLED = True
HTTPCACHE_EXPIRATION_SECS = 86400
HTTPCACHE_DIR = 'httpcache'
HTTPCACHE_IGNORE_HTTP_CODES = []
HTTPCACHE_STORAGE = 'scrapy.extensions.httpcache.FilesystemCacheStorage'

DOWNLOAD_TIMEOUT = 30
DOWNLOAD_WARNSIZE = 0


histdata_scrapy/middlewares.py
# -*- coding: utf-8 -*-

# Define here the models for your spider middleware
#
# See documentation in:
# https://docs.scrapy.org/en/latest/topics/spider-middleware.html

from scrapy import signals


class HistdataScrapySpiderMiddleware(object):
    # Not all methods need to be defined. If a method is not defined,
    # scrapy acts as if the spider middleware does not modify the
    # passed objects.

    @classmethod
    def from_crawler(cls, crawler):
        # This method is used by Scrapy to create your spiders.
        s = cls()
        crawler.signals.connect(s.spider_opened, signal=signals.spider_opened)
        return s

    def process_spider_input(self, response, spider):
        # Called for each response that goes through the spider
        # middleware and into the spider.

        # Should return None or raise an exception.
        return None

    def process_spider_output(self, response, result, spider):
        # Called with the results returned from the Spider, after
        # it has processed the response.

        # Must return an iterable of Request, dict or Item objects.
        for i in result:
            yield i

    def process_spider_exception(self, response, exception, spider):
        # Called when a spider or process_spider_input() method
        # (from other spider middleware) raises an exception.

        # Should return either None or an iterable of Request, dict
        # or Item objects.
        pass

    def process_start_requests(self, start_requests, spider):
        # Called with the start requests of the spider, and works
        # similarly to the process_spider_output() method, except
        # that it doesn’t have a response associated.

        # Must return only requests (not items).
        for r in start_requests:
            yield r

    def spider_opened(self, spider):
        spider.logger.info('Spider opened: %s' % spider.name)


class HistdataScrapyDownloaderMiddleware(object):
    # Not all methods need to be defined. If a method is not defined,
    # scrapy acts as if the downloader middleware does not modify the
    # passed objects.

    @classmethod
    def from_crawler(cls, crawler):
        # This method is used by Scrapy to create your spiders.
        s = cls()
        crawler.signals.connect(s.spider_opened, signal=signals.spider_opened)
        return s

    def process_request(self, request, spider):
        # Called for each request that goes through the downloader
        # middleware.

        # Must either:
        # - return None: continue processing this request
        # - or return a Response object
        # - or return a Request object
        # - or raise IgnoreRequest: process_exception() methods of
        #   installed downloader middleware will be called
        return None

    def process_response(self, request, response, spider):
        # Called with the response returned from the downloader.

        # Must either;
        # - return a Response object
        # - return a Request object
        # - or raise IgnoreRequest
        return response

    def process_exception(self, request, exception, spider):
        # Called when a download handler or a process_request()
        # (from other downloader middleware) raises an exception.

        # Must either:
        # - return None: continue processing this exception
        # - return a Response object: stops process_exception() chain
        # - return a Request object: stops process_exception() chain
        pass

    def spider_opened(self, spider):
        spider.logger.info('Spider opened: %s' % spider.name)


histdata_scrapy/spiders/histdata.py
# -*- coding: utf-8 -*-
import scrapy
from zipfile import ZipFile
from io import BytesIO
import pandas as pd
from sqlalchemy import create_engine
import os


class HistDataSpider(scrapy.Spider):
    name = 'histdata'
    allowed_domains = ['histdata.com']
    start_urls = ['http://www.histdata.com/download-free-forex-data/?/ascii/tick-data-quotes']

    target_fxpairs = ["USD/JPY", "EUR/USD"]
    target_month = (2019, 8)
    # target_month = None

    def parse(self, response):
        self.logger.info(f"parse: url={response.url}")

        for link_cell in response.xpath("//div[@id='content']/div/table/tr/td"):
            url = link_cell.xpath("a/@href").extract()[0]
            url = response.urljoin(url)
            fxpair = link_cell.xpath("a/strong/text()").extract()[0]

            self.logger.info(f"fxpair={fxpair}, url={url}")

            if fxpair not in self.target_fxpairs:
                continue

            yield scrapy.Request(url, callback=self.parse_years, meta={"fxpair": fxpair})

    def parse_years(self, response):
        fxpair = response.meta["fxpair"]
        self.logger.info(f"parse_years: url={response.url}, fxpair={fxpair}")

        for link_cell in response.xpath("//div[@class='page-content']/table/tr/td"):
            url = link_cell.xpath("a/@href").extract()[0]
            url = response.urljoin(url)
            year = int(link_cell.xpath("a/strong/text()").extract()[0])

            self.logger.info(f"year={year}, url={url}")

            if self.target_month is not None:
                if self.target_month[0] != year:
                    continue

            yield scrapy.Request(url, callback=self.parse_months, meta={"fxpair": fxpair, "year": year})

    def parse_months(self, response):
        fxpair = response.meta["fxpair"]
        year = response.meta["year"]
        self.logger.info(f"parse_months: url={response.url}, fxpair={fxpair}, year={year}")

        for link in response.xpath("//div[@class='page-content']/p/a"):
            url = link.xpath("@href").extract()[0]
            url = response.urljoin(url)
            month = int(url.split("/")[-1])

            self.logger.info(f"month={month}, url={url}")

            if self.target_month is not None:
                if self.target_month[0] != year or self.target_month[1] != month:
                    continue

            yield scrapy.Request(url, callback=self.parse_download, meta={"fxpair": fxpair, "year": year, "month": month})

    def parse_download(self, response):
        fxpair = response.meta["fxpair"]
        year = response.meta["year"]
        month = response.meta["month"]
        self.logger.info(f"parse_download: url={response.url}, fxpair={fxpair}, year={year}, month={month}")

        for form_node in response.xpath("//form[@id='file_down']"):
            url = form_node.xpath("@action").extract()[0]
            url = response.urljoin(url)

            params = {
                "tk": form_node.xpath("input[@id='tk']/@value").extract()[0],
                "date": form_node.xpath("input[@id='date']/@value").extract()[0],
                "datemonth": form_node.xpath("input[@id='datemonth']/@value").extract()[0],
                "platform": form_node.xpath("input[@id='platform']/@value").extract()[0],
                "timeframe": form_node.xpath("input[@id='timeframe']/@value").extract()[0],
                "fxpair": form_node.xpath("input[@id='fxpair']/@value").extract()[0]
            }

            self.logger.info(f"url={url}, params={params}")

            yield scrapy.FormRequest(url, formdata=params, callback=self.parse_zip_download, meta={"fxpair": fxpair, "year": year, "month": month})

    def parse_zip_download(self, response):
        fxpair = response.meta["fxpair"]
        year = response.meta["year"]
        month = response.meta["month"]
        self.logger.info(f"parse_zip_download: content-type={response.headers['Content-Type']}, size={len(response.body)}, fxpair={fxpair}, year={year}, month={month}")

        with BytesIO(response.body) as buf:
            with ZipFile(buf) as dl_zip:
                for file_name in dl_zip.namelist():
                    if file_name.endswith(".csv"):
                        with BytesIO(dl_zip.read(file_name)) as csv_buf:
                            df = pd.read_csv(csv_buf, names=["timestamp_str", "bid", "ask", "volume"])
                            df["timestamp"] = pd.to_datetime(df["timestamp_str"] + "000", format="%Y%m%d %H%M%S%f") + pd.Timedelta(hours=5)
                            df = df.set_index("timestamp")

                            self.logger.info(df)

                            for freq in ["10S", "H", "D"]:
                                df_ohlc = df["ask"].resample(freq).ohlc().dropna()
                                df_ohlc["fxpair"] = fxpair
                                df_ohlc["freq"] = freq
                                df_ohlc["timestamp"] = df_ohlc.index
                                df_ohlc["id"] = df_ohlc["fxpair"].str[:3] + df_ohlc["fxpair"].str[-3:] + "_" + df_ohlc["freq"] + "_" + df_ohlc["timestamp"].dt.strftime("%Y%m%d%H%M%S%f")
                                df_ohlc = df_ohlc.set_index("id")

                                self.logger.info(df_ohlc)

                                self.logger.info(f"insert into db start. rows={len(df_ohlc)}")
                                df_ohlc.to_sql(name="historical", con=self.get_db_connection(), if_exists="append", index_label="id")
                                self.logger.info("insert into db finish.")

    def get_db_connection(self):
        url = f"postgresql://{os.environ['DB_USERNAME']}:{os.environ['DB_PASSWORD']}@{os.environ['DB_HOST']}:{os.environ['DB_PORT']}/{os.environ['DB_DATABASE']}"
        return create_engine(url)


db/versions/d47712f990c6_create_ticks.py
"""create ticks

Revision ID: d47712f990c6
Revises:
Create Date: 2019-08-16 17:52:49.530563

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = 'd47712f990c6'
down_revision = None
branch_labels = None
depends_on = None


def upgrade():
    op.create_table(
        "historical",
        sa.Column("id", sa.String(40), primary_key=True),
        sa.Column("timestamp", sa.DateTime, nullable=False),
        sa.Column("fxpair", sa.String(10), nullable=False),
        sa.Column("freq", sa.String(10), nullable=False),
        sa.Column("open", sa.Float, nullable=False),
        sa.Column("high", sa.Float, nullable=False),
        sa.Column("low", sa.Float, nullable=False),
        sa.Column("close", sa.Float, nullable=False),
    )


def downgrade():
    op.drop_table("historical")
